File: mpp-proiect-java-culbec/MotorcycleContest/Model/src/main/java/Model/Entity.java
package Model;

import jakarta.persistence.Id;
import jakarta.persistence.MappedSuperclass;

import java.util.Objects;

@MappedSuperclass
public class Entity<ID> {
    @Id
    private ID id = null;

    Entity() {
    }

    public ID getId() {
        return id;
    }

    public void setId(ID id) {
        this.id = id;
    }

    public static class Builder<ID> {
        public Builder() {
        }

        private Entity<ID> member = new Entity<>();

        public Builder<ID> reset() {
            this.member = new Entity<>();
            return this;
        }
        public Builder<ID> setId(ID id) {
            member.id = id;
            return this;
        }

        public Entity<ID> build() {
            return member;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Entity<?> entity = (Entity<?>) o;
        return Objects.equals(id, entity.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Model.Entity{" +
                "id=" + id +
                '}';
    }
}

File: mpp-proiect-csharp-culbec/Model/Entity.cs
namespace Model;

public class Entity<TId>
{
    public virtual TId? Id { get; set; }

    internal Entity() {}

    public class Builder
    {
        private Entity<TId> _member = new();


        public Builder Reset()
        {
            _member = new Entity<TId>();
            return this;
        }

        public Builder SetId(TId id)
        {
            _member.Id = id;
            return this;
        }

        public Entity<TId> Build()
        {
            return _member;
        }
    }

    public static Builder NewBuilder()
    {
        return new Builder();
    }

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        return obj.GetType() == GetType() && Equals(Id, ((Entity<TId>)obj).Id);
    }

    public override int GetHashCode()
    {
        return EqualityComparer<TId>.Default.GetHashCode(Id!);
    }

    public override string ToString()
    {
        return $"{nameof(Id)}: {Id}";
    }
}